File: Javascript/70.climbing-stairs.js
/*
 * @lc app=leetcode id=70 lang=javascript
 *
 * [70] Climbing Stairs
 */

// @lc code=start
/**
 * @param {number} n
 * @return {number}
 */
var climbStairs = function(n) {
    let step2Count = 0;
    let space = n;
    let step = 1;
    while (space * 2 >= n) {
        step2Count++;
        space = n - step2Count;
        const spaceArray = Array.from({length: space}, (v, k) => k + 1).reverse();
        const step2CountArray = Array.from({length: step2Count}, (v,k) => k + 1);
        if (step2Count < spaceArray.length) {
            const c = spaceArray.slice(0, step2Count).reduce((p, c) => p * c) / step2CountArray.reduce((p,c) => p * c);
            step += c;
        }

        if (step2Count * 2 === n) {
            step++
            return step;
        }
    }

    return step;
};
// @lc code=end

File: Ruby/70.climbing-stairs.rb
#
# @lc app=leetcode id=70 lang=ruby
#
# [70] Climbing Stairs
#

# @lc code=start
# @param {Integer} n
# @return {Integer}
def climb_stairs(n)
    return 1 if n == 0 || n == 1

    table = Array.new(n+1, 0)

    table[0] = 1
    table[1] = 1

    for i in 2...n+1
        table[i] = table[i - 1] + table[i - 2]
    end

    return table[n]
end
# @lc code=end
